synapse/handlers/saml_handler.py

```diff
@@ -133,9 +143,29 @@ def handle_redirect_request(self, client_redirect_url):
             finish_request(request)
             return
 
-        self._auth_handler.complete_sso_login(user_id, request, relay_state)
+        # Complete the interactive auth session or the login.
+        if current_session and current_session.ui_auth_session_id:
+            requester = await self._auth.get_user_by_req(request)
```

Comment:
It seems this is not doing the proper thing here, the request doesn't actually have the proper info in it for `get_user_by_req` to work. This is because it is the SAML redirect.

I think we need to store the user ID in the UI auth info and then some existing code will handle the case of asserting the user has not changed: https://github.com/matrix-org/synapse/blob/392e0c330e8ae47fcccaa55d98163eefc3a73a7f/synapse/handlers/auth.py#L250-L252

Reply:
See 31e570d6b3200459db261e731f45debc561775fe for this.